hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/test/java/com/google/cloud/bigtable/hbase/replication/adapters/IncompatibleMutationAdapterTest.java

```diff
@@ -0,0 +1,238 @@
+import java.nio.charset.StandardCharsets;
+import java.util.*;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hbase.Cell;
+import org.apache.hadoop.hbase.HBaseConfiguration;
+import org.apache.hadoop.hbase.KeyValue;
+import org.apache.hadoop.hbase.client.Connection;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnit;
+import org.mockito.junit.MockitoRule;
+
+@RunWith(JUnit4.class)
+public class IncompatibleMutationAdapterTest {
+
+  class TestIncompatibleMutationAdapter extends IncompatibleMutationAdapter {
+
+    /**
+     * The output for AdaptImpl based on index in the walEntry.getEdit.getCells() collection based
+     * on index in the walEntry.getEdit.getCells() collection.
+     */
+    Map<Integer, List<Cell>> adaptedEntryMap = new HashMap<>();
+
+    /**
+     * The set of indexes where mutations can't be adapted. This set takes precedence over
+     * adaptedEntryMap when same index is set on both.
+     */
+    Set<Integer> incompatibleMutations = new HashSet<>();
+
+    /**
+     * Creates an IncompatibleMutationAdapter with HBase configuration, MetricSource, and CBT Table.
+     *
+     * <p>All subclasses must expose this constructor.
+     *
+     * @param conf HBase configuration. All the configurations required by subclases should come
+     *     from here.
+     * @param metricsExporter Interface for exposing Hadoop metric source.
+     * @param connection CBT table taht is destination of the replicated edits. This
+     */
+    public TestIncompatibleMutationAdapter(
+        Configuration conf, MetricsExporter metricsExporter, Connection connection) {
+      super(conf, metricsExporter, connection);
+    }
+
+    @Override
+    protected List<Cell> adaptIncompatibleMutation(BigtableWALEntry walEntry, int index) {
+      if (incompatibleMutations.contains(index)) {
+        throw new UnsupportedOperationException();
+      }
+      if (!adaptedEntryMap.containsKey(index)) {
+        throw new IllegalStateException("Expected value to be set for index " + index);
+      }
+      return adaptedEntryMap.get(index);
+    }
+
+    public void reset() {
+      incompatibleMutations.clear();
+      adaptedEntryMap.clear();
+    }
+  }
+
+  private static final byte[] rowKey = "rowKey".getBytes(StandardCharsets.UTF_8);
+  private static final byte[] cf = "family".getBytes(StandardCharsets.UTF_8);
+  private static final byte[] qual = "qual".getBytes(StandardCharsets.UTF_8);
+  private static final byte[] val = "value".getBytes(StandardCharsets.UTF_8);
+
+  @Rule public final MockitoRule mockitoRule = MockitoJUnit.rule();
+
+  Configuration conf;
+
+  @Mock Connection connection;
+
+  @Mock MetricsExporter metricsExporter;
+
+  @Mock BigtableWALEntry mockWalEntry;
```

Comment:
Why mock a value class? just instantiate it

Reply:
yes, we were using HBase WAL before and there was no good way to instantiate it. 

Done.